spec/index.md

```diff
@@ -1652,6 +1654,25 @@ There must be at most one call to `ic0.call_on_cleanup` between `ic0.call_new` a
 
     This system call traps if the cycle balance of the canister after transferring cycles decreases below the canister's freezing limit.
 
+-   `ic0.call_cycles_add128_up_to : (max_amount_high : i64, max_amount_low : i64, dst: i32) -> ()`
+
+    This system call moves cycles from the canister balance onto the call under construction, to be transferred with that call.
+
+    This moves the maximum possible amount of cycles onto the call, up to these constraints:
+
+    - It moves no more cycles than represented by a 128-bit value which can be obtained by combining the `max_amount_high` and `max_amount_low` parameters.
+
+    - The cycle balance of the canister after transferring cycles does not decrease below the canister's freezing limit.
```

Comment:
Found an issue while implementing this:
- `call_cycles_add128_up_to(u128::MAX)` puts the canister to just above the freezing limit
- The following `call_perform` tries to charge cycles for sending the message and processing the response. This fails because it would put the canister below the freezing threshold.

We could say that `call_cycles_add128_up_to` moves cycles such that a subsequent `call_perform` does not trap if no extra bytes were added to the message, or we could say that it moves cycles such that a `call_perform` with _any_ amount of bytes would succeed.

Reply:
@dsarlis Could we avoid the issue described above by withdrawing the base fee at ic0.call_new and the argument free at each ic0.call_data_append? In my opinion, this would be cleaner than trying to figure out the amount that must be kept around for ic0.call_perform to succeed.

Reply:
A potential difficulty is what to do if the cycles can't be withdrawn as those other system APIs (ic0.call_new and ic0.call_data_append) don't return any values and thus we could only trap which would be a major breaking change.

Reply:
So saying that "ic0.call_cycles_add128_up_to moves cycles such that a subsequent call_perform does not trap if no extra bytes were added to the message" seems to the best way out after all.

Reply:
> we could say that it moves cycles such that a call_perform with any amount of bytes would succeed.

This might be a good option too. It removes the need to avoid adding more bytes to the call and might be more useful for end users. The extra loss in cycles should be minimal in the grand scheme of things.

Reply:
Wording suggestion:

> `ic0.call_cycles_add128_up_to` moves cycles to the call under construction such that a subsequent `call_perform` does not fail because of an insufficient cycles balance (if no extra bytes were added to the message)

Reply:
Done: https://github.com/dfinity/interface-spec/pull/316/commits/d692008cc315bd8576e2c4522d8afabf9a55af89

Could you please take a look if this commit makes sense to you?

Reply:
Lgtm.